### PR TITLE
deserialize event

### LINKED ISSUE #614: Serializing empty lists and maps
A test event from Xamarin with the latest version resulted in lots of empty maps and lists:

Perhaps this was related to the change in serialization done with some attribute DontSerializeEMpty?

Mainly:
```
	"request": {},
	"fingerprint": [],
	"breadcrumbs": [],
	"extra": {},
	"tags": {}
```
<details>
<summary>Example event</summary>

```
{
	"request": {},
	"contexts": {
		"runtime": {
			"type": "runtime",
			"name": "Mono",
			"version": "6.12.0",
			"raw_description": "Mono 6.12.0 (2020-02/be2226b5a1c)"
		},
		"CurrentUICulture": {
			"Name": "en-US",
			"DisplayName": "English (United States)",
			"Calendar": "GregorianCalendar"
		},
		"CurrentCulture": {
			"Name": "en-US",
			"DisplayName": "English (United States)",
			"Calendar": "GregorianCalendar"
		},
		"device": {
			"type": "device",
			"timezone": "America/Toronto",
			"timezone_display_name": "America/Toronto"
		}
	},
	"user": {},
	"modules": {
		"mscorlib": "2.0.5.0",
		"Xamarin.iOS": "0.0.0.0",
		"System": "2.0.5.0",
		"Mono.Security": "2.0.5.0",
		"Sentry.Samples.Mobile.iOS": "1.0.0.0",
		"Xamarin.Forms.Platform.iOS": "2.0.0.0",
		"Xamarin.Forms.Core": "2.0.0.0",
		"netstandard": "2.1.0.0",
		"System.Core": "2.0.5.0",
		"System.Xml": "2.0.5.0",
		"System.Numerics": "2.0.5.0",
		"System.Data": "2.0.5.0",
		"System.Transactions": "2.0.5.0",
		"System.Data.DataSetExtensions": "2.0.5.0",
		"System.Drawing.Common": "4.0.0.0",
		"System.IO.Compression": "2.0.5.0",
		"System.IO.Compression.FileSystem": "2.0.5.0",
		"System.ComponentModel.Composition": "2.0.5.0",
		"System.Net.Http": "4.0.0.0",
		"System.Runtime.Serialization": "2.0.5.0",
		"System.ServiceModel.Internals": "0.0.0.0",
		"System.Web.Services": "2.0.5.0",
		"System.Xml.Linq": "2.0.5.0",
		"Xamarin.Forms.Platform": "2.0.0.0",
		"Sentry.Samples.Mobile": "1.0.0.0",
		"Xamarin.Forms.Xaml": "2.0.0.0",
		"Sentry": "3.0.0.0",
		"Newtonsoft.Json": "11.0.0.0",
		"System.Threading.Tasks.Extensions": "4.2.0.0",
		"Microsoft.Bcl.AsyncInterfaces": "1.0.0.0",
		"System.Buffers": "4.0.99.0",
		"Xamarin.HotReload.Agent": "1.0.0.0",
		"Xamarin.HotReload.Contracts": "1.0.0.0",
		"Microsoft.VisualStudio.DesignTools.TapContract": "16.0.0.0",
		"Xamarin.HotReload.Forms": "1.0.0.0",
		"Xamarin.HotReload.Translations": "1.0.0.0"
	},
	"event_id": "fc87cdc46d53482dacd9918922cf7be6",
	"timestamp": "2020-11-22T19:57:53.798318+00:00",
	"logentry": {
		"message": "Hello Xamarin.Forms"
	},
	"platform": "csharp",
	"level": "info",
	"environment": "production",
	"sdk": {
		"name": "sentry.dotnet",
		"version": "3.0.0-alpha.4"
	},
	"fingerprint": [],
	"breadcrumbs": [],
	"extra": {},
	"tags": {}
}
```

</details>

### LINKED ISSUE #615: Fix: SentryEvent to get back backing serializable fields
This PR: https://github.com/getsentry/sentry-dotnet/pull/590/files

Broke the deserialization of the fields such as:

```
[DataMember(Name = "tags", EmitDefaultValue = false)]
internal ConcurrentDictionary<string, string>? InternalTags { get; set; }
```

Please rollback to this approach.

### PR DESCRIPTION
**Pull request by bruno-garcia**
Resolves #614
Resolves #615

We need to figure out:

1. Why assigning to SentryException fail to deserialize
2. Deserializing `Contexts` is tricky since it correctly adds to the map, but it's unaware that a well known key such as `browser` should actually deserialize to `Browser`. We need a custom serializer.
3. Complete all Serialization tests into a two-way serialize->deserialize
4. Have snapshot tests to make sure that we know when we break compatibility with already serialized events (that will exist on devices once we ship this).

### CODE REVIEW

File: test/Sentry.Tests/Protocol/Envelopes/EnvelopeTests.cs [hunk 1]
```diff
@@ -350,12 +350,36 @@ public async Task Deserialization_EnvelopeWithItemWithoutLength_Success()
...
+                Level = SentryLevel.Fatal,
+                Logger = "logger",
+                Message = new SentryMessage
+                {
+                    Message = "message",
+                    Formatted = "structured_message"
+                },
+                Modules = { { "module_key", "module_value" } },
+                Release = "release",
+                // SentryExceptions = new[] { new SentryException { Value = "exception_value" } },
```
Comment (bruno-garcia) [inline test/Sentry.Tests/Protocol/Envelopes/EnvelopeTests.cs:372]: This is still broken

**Comment by lucas-zimerman** [comment]
Looks ok for me.

File: src/Sentry/Internal/Json.cs [hunk 2]
```diff
@@ -89,6 +57,7 @@ public static byte[] SerializeToByteArray(object obj)
             return buffer.ToArray();
         }
 
+        [return: MaybeNull]
```
Comment (bruno-garcia) [inline src/Sentry/Internal/Json.cs:60]: bumped Newtonsoft.Json to 12.0.3 and this was required

File: src/Sentry/Protocol/Breadcrumb.cs [hunk 3]
```diff
@@ -15,7 +16,7 @@ namespace Sentry.Protocol
     public sealed class Breadcrumb
     {
         [DataMember(Name = "timestamp", EmitDefaultValue = false)]
-        private string SerializableTimestamp => Timestamp.ToString("yyyy-MM-ddTHH\\:mm\\:ssZ", DateTimeFormatInfo.InvariantInfo);
+        private string SerializableTimestamp => Timestamp.ToString("yyyy-MM-ddTHH\\:mm\\:ss.fffZ", DateTimeFormatInfo.InvariantInfo);
```
Comment (bruno-garcia) [inline src/Sentry/Protocol/Breadcrumb.cs:19]: Improved breadcrumb precision in the process

File: src/Sentry/Protocol/SentryEvent.cs [hunk 4]
```diff
@@ -167,25 +165,32 @@ public User User
         [DataMember(Name = "sdk", EmitDefaultValue = false)]
         public SdkVersion Sdk { get; internal set; } = new SdkVersion();
 
-        /// <inheritdoc />
         [DataMember(Name = "fingerprint", EmitDefaultValue = false)]
-        [DontSerializeEmpty]
-        public IEnumerable<string> Fingerprint { get; set; } = Enumerable.Empty<string>();
+        private IEnumerable<string>? _fingerprint;
```
Comment (bruno-garcia) [inline src/Sentry/Protocol/SentryEvent.cs:169]: Got rid of the attribute which seems not to work (at least on Mono)

File: src/Sentry/Protocol/SentryEvent.cs [hunk 5]
```diff
@@ -167,25 +165,32 @@ public User User
...
         [DataMember(Name = "breadcrumbs", EmitDefaultValue = false)]
-        [DontSerializeEmpty]
-        public IEnumerable<Breadcrumb> Breadcrumbs { get; } = new List<Breadcrumb>();
+        private List<Breadcrumb>? _breadcrumbs;
 
         /// <inheritdoc />
-        [DataMember(Name = "extra", EmitDefaultValue = false)]
-        [DontSerializeEmpty]
-        public IReadOnlyDictionary<string, object?> Extra { get; } = new Dictionary<string, object?>();
+        public IEnumerable<Breadcrumb> Breadcrumbs => _breadcrumbs ??= new List<Breadcrumb>();
```
Comment (bruno-garcia) [inline src/Sentry/Protocol/SentryEvent.cs:183]: We don't need to new-up a list unless there's access to the property

File: src/Sentry/Sentry.csproj [hunk 6]
```diff
@@ -18,7 +18,7 @@
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="Newtonsoft.Json" Version="11.0.2" />
+    <PackageReference Include="Newtonsoft.Json" Version="12.0.3" />
```
Comment (bruno-garcia) [inline src/Sentry/Sentry.csproj:21]: An attempt to figure out what's going on

File: test/Sentry.Tests/Protocol/SentryEventTests.cs [hunk 7]
```diff
@@ -29,111 +39,41 @@ public void SerializeObject_AllPropertiesSetToNonDefault_SerializesValidObject()
                 },
                 Modules = { { "module_key", "module_value" } },
                 Release = "release",
-                SentryExceptions = new[] { new SentryException { Value = "exception_value" } },
+                // SentryExceptions = new[] { new SentryException { Value = "exception_value" } },
```
Comment (bruno-garcia) [inline test/Sentry.Tests/Protocol/SentryEventTests.cs:42]: Setting this breaks the test with:

`System.Runtime.Serialization.SerializationException: Member 'Message' was not found.`

Even if simply setting an empty list! :gofigure:

**Review by bruno-garcia — COMMENTED** [review]
The protocol can't deserialize.

If assigning a `SentryException` to the event, even if simply an empty array, it fails with:

```
Sentry.Tests.Protocol.SentryEventTests.SerializeObject_AllPropertiesSetToNonDefault_SerializesValidObject

System.Runtime.Serialization.SerializationException: Member 'Message' was not found.

System.Runtime.Serialization.SerializationException
Member 'Message' was not found.
   at System.Runtime.Serialization.SerializationInfo.GetElement(String name, Type& foundType)
   at System.Runtime.Serialization.SerializationInfo.GetString(String name)
   at System.Exception..ctor(SerializationInfo info, StreamingContext context)
   at lambda_method124(Closure , Object[] )
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateISerializable(JsonReader reader, JsonISerializableContract contract, JsonProperty member, String id)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObject(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateValueInternal(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.ResolvePropertyAndCreatorValues(JsonObjectContract contract, JsonProperty containerProperty, JsonReader reader, Type objectType)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObjectUsingCreatorWithParameters(JsonReader reader, JsonObjectContract contract, JsonProperty containerProperty, ObjectConstructor`1 creator, String id)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateNewObject(JsonReader reader, JsonObjectContract objectContract, JsonProperty containerMember, JsonProperty containerProperty, String id, Boolean& createdFromNonDefaultCreator)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObject(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateValueInternal(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, Boolean checkAdditionalContent)
   at Newtonsoft.Json.JsonSerializer.DeserializeInternal(JsonReader reader, Type objectType)
   at Newtonsoft.Json.JsonSerializer.Deserialize(JsonReader reader, Type objectType)
   at Newtonsoft.Json.JsonConvert.DeserializeObject(String value, Type type, JsonSerializerSettings settings)
   at Newtonsoft.Json.JsonConvert.DeserializeObject[T](String value, JsonSerializerSettings settings)
   at Newtonsoft.Json.JsonConvert.DeserializeObject[T](String value)
   at Sentry.Tests.Protocol.SentryEventTests.SerializeObject_AllPropertiesSetToNonDefault_SerializesValidObject() in /Users/bruno/git/sentry-dotnet/test/Sentry.Tests/Protocol/SentryEventTests.cs:line 64

```

**Review by Tyrrrz — APPROVED** [review]

**Comment by Tyrrrz** [comment]
@bruno-garcia Can be closed

**Pull request closed without merging**